File: indexer-go/internal/combat/xp.go
package combat

// XP math (plan §3.4), mirroring GameLogic._calcXp / settle_battle::calc_xp
// and PetCore.recordBattleOpponent / PetAccount::record_battle_opponent.
// Validated against contracts/test-vectors/xp.json.

// calcXP returns the XP credited for a battle:
//
//	baseXp * clamp(100 + 10*(oppLevel - myLevel), 0, 200) / 100
//
// Punching up (+10 levels) pays double; fighting 10+ levels down pays zero.
func calcXP(baseXP uint32, myLevel, oppLevel uint16) uint32 {
	diff := int32(oppLevel) - int32(myLevel)
	mult := 100 + 10*diff
	if mult <= 0 {
		return 0
	}
	if mult > 200 {
		mult = 200
	}
	return baseXP * uint32(mult) / 100
}

// DecayShift is the same-opponent decay exponent for the nth consecutive
// battle against the same opponent (plan §3.4): XP is right-shifted by this,
// halving each repeat (100 → 50 → 25 → …). Switching opponents resets it.
//
// Given a pet whose lastOpponentId/streak start at 0 (a fresh pet), applyDecay
// folds an opponent sequence and reports the shift applied at each step.
func applyDecay(opponentIDs []uint32) []uint32 {
	shifts := make([]uint32, len(opponentIDs))
	var lastOpponent uint32 // 0 = no battles yet (next_pet_id starts at 1)
	var streak uint32
	for i, opp := range opponentIDs {
		if opp == lastOpponent && lastOpponent != 0 {
			streak++
		} else {
			streak = 0
		}
		shifts[i] = streak
		lastOpponent = opp
	}
	return shifts
}

// Level-up progression, ported from PetCore.addXp / PetAccount::add_xp so Go can
// independently recompute a full progression delta, not just the XP formula and
// decay (plan §3.4). Mirrors protocol/src/combat/xp.ts's applyXp exactly,
// including the two behaviours easiest to get subtly wrong:
//
//   - a pet at the level cap accrues nothing at all, not capped XP: the on-chain
//     version returns before touching xp, so this does too;
//   - at most one level per battle, with the remainder carried as XP rather than
//     the threshold being reapplied.
const XPPerLevelMultiplier = 100

// MaxSameOpponentStreak: sameOpponentStreak is a uint8 on both chains and
// saturates rather than wraps.
const MaxSameOpponentStreak = 255

// MaxDecayShift caps the right-shift applied to an XP award.
//
// The streak can reach 255, but the XP being shifted fits in uint32. Go's `>>`
// on a fixed-width unsigned integer already yields 0 once the shift reaches or
// exceeds the operand's bit width (unlike JavaScript's `>>`, which masks the
// shift count to 5 bits), so this constant exists to document the same ceiling
// the TypeScript port has to enforce explicitly, not to work around a Go
// quirk — go vet would catch a shift literal outside this range, but a
// runtime-computed shift needs the same clamp as every other port.
const MaxDecayShift = 31

// ApplyDecayShift applies same-opponent decay to an XP award.
func ApplyDecayShift(xp uint32, decayShift uint32) uint32 {
	return xp >> min(decayShift, MaxDecayShift)
}

// OpponentHistory is a pet's same-opponent tracking state, as frozen in a
// snapshot. Mirrors protocol's OpponentHistory.
type OpponentHistory struct {
	LastOpponentID uint64 // pet id; 0 = no battles yet
	Streak         uint32
}

// OpponentHistoryUpdate is OpponentHistory after a battle, plus the shift that
// battle earned.
type OpponentHistoryUpdate struct {
	OpponentHistory
	DecayShift uint32
}

// RecordBattleOpponent advances a pet's same-opponent history, mirroring
// recordBattleOpponent / record_battle_opponent. Fighting the same opponent
// again increments the streak (saturating at 255); the new value is the shift,
// so the second consecutive rematch pays half, the third a quarter. Facing
// anyone else resets to 0.
func RecordBattleOpponent(history OpponentHistory, opponentID uint64) OpponentHistoryUpdate {
	if history.LastOpponentID == opponentID {
		streak := history.Streak
		if streak < MaxSameOpponentStreak {
			streak++
		}
		return OpponentHistoryUpdate{
			OpponentHistory: OpponentHistory{LastOpponentID: history.LastOpponentID, Streak: streak},
			DecayShift:      streak,
		}
	}
	return OpponentHistoryUpdate{
		OpponentHistory: OpponentHistory{LastOpponentID: opponentID, Streak: 0},
		DecayShift:      0,
	}
}

// LevelState is a pet's level and XP.
type LevelState struct {
	Level uint16
	XP    uint32
}

// LevelStateUpdate is LevelState after an XP award.
type LevelStateUpdate struct {
	LevelState
	LeveledUp bool
}

// ApplyXP credits XP and advances at most one level, mirroring
// PetCore.addXp / PetAccount::add_xp.
func ApplyXP(state LevelState, amount uint32, maxLevel uint16) LevelStateUpdate {
	if state.Level >= maxLevel {
		return LevelStateUpdate{LevelState: state, LeveledUp: false}
	}
	xp := state.XP + amount
	level := state.Level
	threshold := uint32(XPPerLevelMultiplier) * uint32(level)
	leveledUp := false
	if xp >= threshold {
		xp -= threshold
		level++
		if level > maxLevel {
			level = maxLevel
		}
		leveledUp = true
	}
	return LevelStateUpdate{LevelState: LevelState{Level: level, XP: xp}, LeveledUp: leveledUp}
}

File: protocol/src/combat/xp.ts
/**
 * XP and level progression, ported from the on-chain implementations:
 * `GameLogic._calcXp` + `PetCore.addXp` + `PetCore.recordBattleOpponent` (Solidity),
 * `game::xp::calc_xp` + `PetAccount::add_xp` + `PetAccount::record_battle_opponent`
 * (Rust), and `indexer-go/internal/combat/xp.go` (Go, which covers the formula and
 * the decay but not level-up).
 *
 * Validated against `contracts/test-vectors/xp.json`, the same file Hardhat, Anchor,
 * and indexer-go consume. If a case fails, this port drifted; fix the port, never
 * the vector.
 *
 * Pure number math, no snapshot awareness. The snapshot-shaped wrapper that turns a
 * battle result into a progression delta lives in `src/progression/`, so this file
 * stays a line-for-line analogue of its siblings.
 */

/** Base XP for the winner, before the level multiplier and decay. */
export const BASE_XP_WIN = 100;
/** Base XP for the loser. */
export const BASE_XP_LOSS = 25;
/** XP needed to advance: `100 * currentLevel`. */
export const XP_PER_LEVEL_MULTIPLIER = 100;
/** `sameOpponentStreak` is a uint8 on both chains and saturates rather than wrapping. */
export const MAX_SAME_OPPONENT_STREAK = 255;
/**
 * Ceiling on the decay shift.
 *
 * The streak can reach 255, but the XP being shifted is a uint32. Solidity defines a
 * shift at or past the operand width as 0; Rust would panic with overflow checks on,
 * so `settle_battle.rs` clamps with `.min(31)`; Go also yields 0. JavaScript is the
 * odd one out: `>>` masks the shift count to 5 bits, so `200 >> 32` is `200`, not 0.
 * Clamping here is what keeps this port from silently paying full XP exactly where
 * the chains pay none.
 */
export const MAX_DECAY_SHIFT = 31;

/** Default level cap, mirroring `GameConfig.maxLevel`'s initializer. Owner-tunable on chain. */
export const DEFAULT_MAX_LEVEL = 100;

/**
 * XP for one battle before decay:
 * `baseXp * clamp(100 + 10 * (oppLevel - myLevel), 0, 200) / 100`.
 *
 * Punching up ten levels pays double; fighting ten levels down pays nothing.
 */
export function calcXp(baseXp: number, myLevel: number, oppLevel: number): number {
    const diff = oppLevel - myLevel;
    const mult = 100 + 10 * diff;
    if (mult <= 0) {
        return 0;
    }
    const capped = mult > 200 ? 200 : mult;
    return Math.floor((baseXp * capped) / 100);
}

/** Applies same-opponent decay to an XP award. */
export function applyDecayShift(xp: number, decayShift: number): number {
    const shift = decayShift > MAX_DECAY_SHIFT ? MAX_DECAY_SHIFT : decayShift;
    return xp >>> shift;
}

/** A pet's same-opponent tracking state, as frozen in a snapshot. */
export interface OpponentHistory {
    /** Previous opponent, or 0 for a pet that has not fought. */
    lastOpponentId: bigint;
    /** Consecutive prior battles against `lastOpponentId`. */
    streak: number;
}

/** `OpponentHistory` after a battle, plus the shift that battle earned. */
export interface OpponentHistoryUpdate extends OpponentHistory {
    /** XP right-shift for this battle: 0 = full, 1 = half, 2 = quarter. */
    decayShift: number;
}

/**
 * Advances a pet's same-opponent history, mirroring `recordBattleOpponent`.
 *
 * Fighting the same opponent again increments the streak (saturating at 255) and the
 * new value is the shift, so the second consecutive rematch pays half, the third a
 * quarter. Facing anyone else resets to 0, which is why grinding one opponent stops
 * being worth it while switching targets always pays full.
 */
export function recordBattleOpponent(history: OpponentHistory, opponentId: bigint): OpponentHistoryUpdate {
    if (history.lastOpponentId === opponentId) {
        const streak =
            history.streak < MAX_SAME_OPPONENT_STREAK ? history.streak + 1 : MAX_SAME_OPPONENT_STREAK;
        return { lastOpponentId: history.lastOpponentId, streak, decayShift: streak };
    }
    return { lastOpponentId: opponentId, streak: 0, decayShift: 0 };
}

/** A pet's level and XP. */
export interface LevelState {
    level: number;
    xp: number;
}

/** `LevelState` after an XP award. */
export interface LevelStateUpdate extends LevelState {
    leveledUp: boolean;
}

/**
 * Credits XP and advances at most one level, mirroring `PetCore.addXp` /
 * `PetAccount::add_xp`.
 *
 * Three behaviours worth being explicit about, because all three are easy to
 * "improve" into a divergence from the chains:
 *
 * - A pet at the level cap accrues nothing at all. Not capped XP, none: the on-chain
 *   version returns before touching `xp`.
 * - At most one level per battle. Leftover XP beyond a second threshold stays as XP.
 * - The threshold is `100 * level` at the pre-increment level.
 */
export function applyXp(state: LevelState, amount: number, maxLevel: number = DEFAULT_MAX_LEVEL): LevelStateUpdate {
    if (state.level >= maxLevel) {
        return { level: state.level, xp: state.xp, leveledUp: false };
    }
    let xp = state.xp + amount;
    let level = state.level;
    const threshold = XP_PER_LEVEL_MULTIPLIER * level;
    let leveledUp = false;
    if (xp >= threshold) {
        xp -= threshold;
        level += 1;
        if (level > maxLevel) {
            level = maxLevel;
        }
        leveledUp = true;
    }
    return { level, xp, leveledUp };
}
